plugin/conn_ip_example/conn_ip_example_test.go

```diff
@@ -11,46 +11,122 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package main_test
+package main
 
 import (
 	"context"
+	"fmt"
+	"strconv"
+	"sync/atomic"
 	"testing"
 
 	. "github.com/pingcap/check"
+	"github.com/pingcap/log"
 	"github.com/pingcap/tidb/plugin"
+	"github.com/pingcap/tidb/sessionctx/variable"
 )
 
-func LoadRunShutdownPluginExample() {
+type testConnIPExampleSuite struct{}
+
+var _ = SerialSuites(&testConnIPExampleSuite{})
+
+func (s *testConnIPExampleSuite) TestLoadPlugin(c *C) {
 	ctx := context.Background()
-	var pluginVarNames []string
+	pluginName := "conn_ip_example"
+	pluginVersion := uint16(1)
+	pluginSign := pluginName + "-" + strconv.Itoa(int(pluginVersion))
+
 	cfg := plugin.Config{
-		Plugins:        []string{"conn_ip_example-1"},
-		PluginDir:      "/home/robi/Code/go/src/github.com/pingcap/tidb/plugin/conn_ip_example",
-		PluginVarNames: &pluginVarNames,
+		Plugins:        []string{pluginSign},
+		PluginDir:      "",
+		PluginVarNames: &variable.PluginVarNames,
+		EnvVersion:     map[string]uint16{"go": 1112},
+	}
+
+	// setup load test hook.
+	loadOne := func(p *plugin.Plugin, dir string, pluginID plugin.ID) (manifest func() *plugin.Manifest, err error) {
+		return func() *plugin.Manifest {
+			m := &plugin.AuditManifest{
+				Manifest: plugin.Manifest{
+					Kind:       plugin.Audit,
+					Name:       pluginName,
+					Version:    pluginVersion,
+					SysVars:    map[string]*variable.SysVar{pluginName + "_key": {Scope: variable.ScopeGlobal, Name: pluginName + "_key", Value: "v1"}},
+					OnInit:     OnInit,
+					OnShutdown: OnShutdown,
+					Validate:   Validate,
+				},
+				OnGeneralEvent:    OnGeneralEvent,
+				OnConnectionEvent: OnConnectionEvent,
+			}
+			return plugin.ExportManifest(m)
+		}, nil
 	}
+	plugin.SetTestHook(loadOne)
 
+	// trigger load.
 	err := plugin.Load(ctx, cfg)
 	if err != nil {
-		panic(err)
+		log.Fatal(fmt.Sprintf("load plugin [%s] fail, error [%s]\n", pluginSign, err))
 	}
 
-	// load and start TiDB domain.
-
 	err = plugin.Init(ctx, cfg)
 	if err != nil {
-		panic(err)
+		log.Fatal(fmt.Sprintf("init plugin [%s] fail, error [%s]\n", pluginSign, err))
 	}
 
 	err = plugin.ForeachPlugin(plugin.Audit, func(auditPlugin *plugin.Plugin) error {
```

Comment:
@xhebox  the static check of tidb is too strict... I think variable err must be nil here.

Reply:
Then this function should not return error. Or we can add this function into the [errcheck_excludes.txt](https://github.com/pingcap/tidb/blob/master/tools/check/errcheck_excludes.txt) after some investigations.